File: typescript/src/collect.ts
/**
 * Read a request body into a record. Pure, no network, no tracer.
 *
 * AMBIGUITY 3 (now fixed in the spec): the spec never said how a `parsed` payload is serialised before hashing, and it
 * cannot -- the first implementation used Python's `repr`, which no other language reproduces. `0.0` renders as "0.0"
 * there and "0" here, so a `decoding` digest computed by these two implementations of the same request will differ.
 *
 * That is now stated rather than papered over: a `parsed` digest is **collector-local**. It is comparable only against
 * another digest from the same collector and version, and since a `parsed` digest can never key an identity, nothing
 * that matters was lost. What WAS lost before saying it: somebody would have compared two and concluded the harnesses
 * differed.
 *
 * AMBIGUITY 4 (now fixed in the spec): with several system messages, nothing said how they combine. Joining with a
 * newline is what the first implementation did, and because the instruction digest KEYS THE IDENTITY, two collectors
 * choosing differently would give two identities for one request.
 */
import { Absence, Manifest, Part, Record_, digest } from "./record.js";
import { PARTS } from "./vocab.js";

/** Static per version: a manifest that adapted to what it found could never contradict a record. */
export const FROM_A_REQUEST = ["instruction", "tool_schemas", "readout", "decoding"] as const;

/** Only the instruction is text the model provably read; the rest are protocol values whose rendering we do not hold. */
export const REQUEST_BOUNDARIES: Record<string, string> = {
  instruction: "model_visible",
  tool_schemas: "parsed",
  readout: "parsed",
  decoding: "parsed",
};

export const DECODING_KEYS = ["temperature", "top_p", "top_k", "seed"] as const;

/** Parts a request body has no place to put. Their absence is the SENDER's. */
export const PUSHED_PARTS = ["loop", "turn_budget", "retry_policy", "context_partitioning"] as const;

/** Collector-local rendering of protocol scalars. See AMBIGUITY 3. */
function scalars(pairs: [string, unknown][]): string {
  return pairs.map(([k, v]) => `${k}=${JSON.stringify(v)}`).join(";");
}

export interface Body {
  messages?: { role?: string; content?: unknown }[];
  tools?: { function?: { name?: string; parameters?: unknown }; name?: string; parameters?: unknown }[];
  response_format?: unknown;
  [key: string]: unknown;
}

export function fromRequest(body: Body, opts: { collector: string; version: string }): Record_ {
  const parts: Part[] = [];
  const absences: Absence[] = [];

  const hold = (kind: string, payload: string, label = ""): void => {
    parts.push(
      new Part({
        kind,
        sourcing: "in_the_request",
        label,
        contentDigest: digest(payload),
        boundary: REQUEST_BOUNDARIES[kind] as string,
      }),
    );
  };

  const messages = body.messages ?? [];
  // Newline-joined, per the spec's serialisation section. This keys the identity, so it is normative rather than local.
  const instruction = messages
    .filter((m) => m.role === "system")
    .map((m) => String(m.content ?? ""))
    .join("\n");
  if (instruction.trim() !== "") {
    // Hashed RAW, down to case and whitespace: the model reads the string and not its meaning.
    hold("instruction", instruction);
  } else {
    absences.push(new Absence("instruction", "not_provided"));
  }

  const tools = body.tools;
  if (tools !== undefined && tools.length > 0) {
    // Sorted: the order a caller lists tools in is not part of what it offered.
    const shapes = tools
      .map((t) => {
        const fn = t.function ?? t;
        return scalars([
          ["name", fn.name],
          ["params", fn.parameters],
        ]);
      })
      .sort();
    hold("tool_schemas", shapes.join("|"), `${tools.length} tool(s)`);
  } else {
    absences.push(new Absence("tool_schemas", "not_provided"));
  }

  const fmt = body.response_format;
  if (fmt !== undefined && fmt !== null) {
    const label = typeof fmt === "object" && fmt !== null && "type" in fmt ? String((fmt as { type: unknown }).type) : "";
    hold("readout", scalars([["response_format", fmt]]), label);
  } else {
    absences.push(new Absence("readout", "not_provided"));
  }

  const knobs = DECODING_KEYS.filter((k) => k in body).map((k) => [k, body[k]] as [string, unknown]);
  if (knobs.length > 0) {
    hold("decoding", scalars(knobs), knobs.map(([k]) => k).join(","));
  } else {
    absences.push(new Absence("decoding", "not_provided"));
  }

  for (const kind of PUSHED_PARTS) absences.push(new Absence(kind, "not_provided"));
  absences.push(new Absence("tool_extension", "not_observable"));
  // At request time the calls have not happened, so there is nothing the sender could have supplied.
  absences.push(new Absence("tool_trace", "not_provided"));

  const record = new Record_(
    new Manifest(opts.collector, opts.version, [...FROM_A_REQUEST]),
    "complete",
    parts,
    absences,
  );
  if (record.unaccounted.length > 0) {
    throw new Error(`the collector left ${record.unaccounted} unaccounted for`);
  }
  return record;
}

/** Attach what the owner says about the parts a request cannot carry, replacing their `not_provided` absences. */
export function declare(record: Record_, pushed: Partial<Record<(typeof PUSHED_PARTS)[number], string>>): Record_ {
  const unknown = Object.keys(pushed).filter((k) => !(PUSHED_PARTS as readonly string[]).includes(k));
  if (unknown.length > 0) {
    throw new Error(
      `${unknown} are not parts a request body fails to carry; a declaration about something we can read from the ` +
        `request would be a claim standing in front of the evidence`,
    );
  }
  const kept = record.absences.filter((a) => !(a.kind in pushed));
  const declared = Object.entries(pushed)
    .sort(([a], [b]) => (a < b ? -1 : 1))
    .map(
      ([kind, value]) =>
        new Part({
          kind,
          sourcing: "pushed_by_owner",
          label: value as string,
          contentDigest: digest(value as string),
          boundary: "parsed",
        }),
    );
  return new Record_(record.manifest, record.status, [...record.parts, ...declared], kept);
}

export { PARTS };

File: src/perigraph/collect.py
"""Read a request body into a record. Pure, and that is the whole design of it.

No network and no tracer: a shim installed into a sender's process should not be able to fail because something it did
not need moved. It takes the dict a client is about to send.

**What it can and cannot reach, and why that is the finding rather than a limitation.** Only the instruction is text the
model provably read. A `tools` array, a `response_format` and the sampler settings are protocol values: the provider
renders them into the prompt however it likes, or not at all, and the sender does not hold that rendering. So they are
recorded with a `parsed` boundary and cannot key an identity -- which means **a harness read from a request body is
identified by its instruction and nothing else.**

That is the same fact the measurement found from the other side: one sentence of the instruction moved accuracy 12.04
points, and it is also the only part of a request whose exact bytes we can prove the model read.
"""
from __future__ import annotations

from perigraph.record import Absence, Manifest, Part, Record, digest
from perigraph.vocab import PARTS

#: What this collector claims it can reach from a request body, and nothing more. **Static per version on purpose:** a
#: manifest that adapted to what it happened to find could never contradict a record, which is the one thing it is for.
FROM_A_REQUEST = ("instruction", "tool_schemas", "readout", "decoding")

#: Which boundary each of those digests is actually over. This table is the finding; see the module docstring.
REQUEST_BOUNDARIES = {
    "instruction": "model_visible",
    "tool_schemas": "parsed",
    "readout": "parsed",
    "decoding": "parsed",
}

#: Sampler settings read from a body, in a fixed order. Fixed rather than sorted at use, so that adding a knob is a
#: visible edit here instead of a silent change to every digest ever computed.
DECODING_KEYS = ("temperature", "top_p", "top_k", "seed")

#: Parts a request body has no place to put. Their absence is the SENDER's -- the shim did not fail to find them, there
#: is nowhere in a request for them to be -- and saying so is the difference the absence vocabulary exists for.
PUSHED_PARTS = ("loop", "turn_budget", "retry_policy", "context_partitioning")


def _scalars(pairs: list[tuple[str, object]]) -> str:
    """Render protocol scalars for hashing, in the order given.

    Canonicalising **is** allowed here, and the rule that forbids it for the instruction says why: a machine reads these,
    so two spellings of the same setting behave identically. The instruction is the opposite case and is hashed raw.
    """
    return ";".join(f"{k}={v!r}" for k, v in pairs)


def from_request(body: dict, *, collector: str, version: str) -> Record:
    """Read what a request body says about the harness, and record whose absence each hole is."""
    parts: list[Part] = []
    absences: list[Absence] = []

    def hold(kind: str, payload: str, label: str = "") -> None:
        parts.append(Part(kind=kind, sourcing="in_the_request", label=label,
                          content_digest=digest(payload), boundary=REQUEST_BOUNDARIES[kind]))

    messages = body.get("messages") or []
    instruction = "\n".join(str(m.get("content", "")) for m in messages if m.get("role") == "system")
    if instruction.strip():
        # Hashed RAW, down to case and whitespace. The model reads the string and not its meaning, and 0.6243 against
        # 0.7447 came from one sentence's wording -- so two spellings are two harnesses even where a human would call
        # them the same instruction.
        hold("instruction", instruction)
    else:
        absences.append(Absence(kind="instruction", reason="not_provided"))

    tools = body.get("tools")
    if tools:
        # Sorted by rendered shape: the order a caller lists tools in is not part of what it offered, and an unsorted
        # digest would make two identical toolsets look like two harnesses.
        shapes = sorted(_scalars([("name", (t.get("function") or t).get("name")),
                                 ("params", (t.get("function") or t).get("parameters"))]) for t in tools)
        hold("tool_schemas", "|".join(shapes), label=f"{len(tools)} tool(s)")
    else:
        absences.append(Absence(kind="tool_schemas", reason="not_provided"))

    fmt = body.get("response_format")
    if fmt:
        hold("readout", _scalars([("response_format", fmt)]),
             label=str(fmt.get("type", "")) if isinstance(fmt, dict) else "")
    else:
        # Not our failure. Without a declared format the answer is read out of prose by a convention on the caller's
        # side, and a mis-set convention has been measured to put 1,822 of 2,364 answers on one option.
        absences.append(Absence(kind="readout", reason="not_provided"))

    knobs = [(k, body[k]) for k in DECODING_KEYS if k in body]
    if knobs:
        hold("decoding", _scalars(knobs), label=",".join(k for k, _ in knobs))
    else:
        absences.append(Absence(kind="decoding", reason="not_provided"))

    for kind in PUSHED_PARTS:
        absences.append(Absence(kind=kind, reason="not_provided"))
    absences.append(Absence(kind="tool_extension", reason="not_observable"))
    # At request time the calls have not happened, so there is nothing the sender could have supplied.
    absences.append(Absence(kind="tool_trace", reason="not_provided"))

    record = Record(manifest=Manifest(collector=collector, version=version, reaches=FROM_A_REQUEST),
                    status="complete", parts=tuple(parts), absences=tuple(absences))
    assert not record.unaccounted, f"the collector left {record.unaccounted} unaccounted for"
    return record


def declare(record: Record, **pushed: str) -> Record:
    """Attach what the owner says about the parts a request cannot carry, replacing their `not_provided` absences.

    Separate from `from_request` because the two are different kinds of fact and the record has to keep them apart. What
    a caller declares about its own loop is a claim we cannot check, so it enters as `pushed_by_owner` and can never key
    an identity however true it happens to be.
    """
    unknown = sorted(set(pushed) - set(PUSHED_PARTS))
    if unknown:
        raise ValueError(f"{unknown} are not parts a request body fails to carry; a declaration about something we can "
                         f"read from the request would be a claim standing in front of the evidence")
    kept = tuple(a for a in record.absences if a.kind not in pushed)
    declared = tuple(Part(kind=k, sourcing="pushed_by_owner", label=v, content_digest=digest(v), boundary="parsed")
                     for k, v in sorted(pushed.items()))
    return Record(manifest=record.manifest, status=record.status,
                  parts=record.parts + declared, absences=kept)
